### ISSUE TITLE
Distilling a T5 model: Need appropriate values for model_parallelism and tokens_per_batch to train a t5.small_ssm model on v3_512, v3_1024 and v3_2048 TPUs

### ISSUE THREAD

**Issue by sbhaktha** [google-research/text-to-text-transfer-transformer#517]
Hello,

I cannot figure out appropriate values for the `model_parallelism` and `tokens_per_batch` parameters to train a t5 small_ssm model. Specifically, what I am trying to do is distill from an `xxl_ssm` teacher model that was fine-tuned on Natural Questions, to a `small_ssm` student model. I need to fit both the teacher and student models in memory and I attempted to do this training on a v3-256 but got OOM issues. I tried several different combinations of the parameters on v3-512 and v3-1024 TPUs but couldn't get any to work. My command line currently looks like this-

```
python -m t5.models.mesh_transformer_main \
    --tpu=<v3_1024_tpu> \
    --gcp_project=<...> \
    --tpu_zone=<...> \
    --model_dir=<...> \
    --gin_file="dataset.gin" \
    --gin_file="distillation_nqopen_xxl_ssm_to_small_ssm_operative_config.gin" \
    --gin_param="MIXTURE_NAME = <...>" \
    --gin_param="utils.run.save_checkpoints_steps=<...>" \
    --gin_param="utils.run.batch_size=('tokens_per_batch', 393216)" \
    --gin_param="utils.run.train_steps=<...> \
    --gin_param="constant_learning_rate.learning_rate=<...>" \
    --gin_param="utils.tpu_mesh_shape.tpu_topology='16x32'" \
    --gin_param="utils.tpu_mesh_shape.model_parallelism=8"   \
```

The distillation gin file is attached.
[distillation_nqopen_xxl_ssm_to_small_ssm_operative_config.gin.txt](https://github.com/tensorflow/mesh/files/5514942/distillation_nqopen_xxl_ssm_to_small_ssm_operative_config.gin.txt)

Running the above script on a `v3-1024` with the attached `gin` file, I get the error:

```
WARNING:tensorflow:Reraising captured error
W1110 05:11:05.038959 140634039817600 error_handling.py:149] Reraising captured error
Traceback (most recent call last):
  File "/home/sumithrab/miniconda/envs/t2t/lib/python3.7/runpy.py", line 193, in _run_module_as_main
    "__main__", mod_spec)
  File "/home/sumithrab/miniconda/envs/t2t/lib/python3.7/runpy.py", line 85, in _run_code
    exec(code, run_globals)
  File "/home/sumithrab/miniconda/envs/t2t/lib/python3.7/site-packages/t5/models/mesh_transformer_main.py", line 264, in <module>
    console_entry_point()
  File "/home/sumithrab/miniconda/envs/t2t/lib/python3.7/site-packages/t5/models/mesh_transformer_main.py", line 261, in console_en
try_point
    app.run(main)
  File "/home/sumithrab/miniconda/envs/t2t/lib/python3.7/site-packages/absl/app.py", line 303, in run
    _run_main(main, args)
  File "/home/sumithrab/miniconda/envs/t2t/lib/python3.7/site-packages/absl/app.py", line 251, in _run_main  
    sys.exit(main(argv))
  File "/home/sumithrab/miniconda/envs/t2t/lib/python3.7/site-packages/t5/models/mesh_transformer_main.py", line 255, in main
    model_dir=FLAGS.model_dir)
  File "/home/sumithrab/miniconda/envs/t2t/lib/python3.7/site-packages/gin/config.py", line 1078, in gin_wrapper
    utils.augment_exception_message_and_reraise(e, err_str)
  File "/home/sumithrab/miniconda/envs/t2t/lib/python3.7/site-packages/gin/utils.py", line 49, in augment_exception_message_and_rer
aise
    six.raise_from(proxy.with_traceback(exception.__traceback__), None)
  File "<string>", line 3, in raise_from
  File "/home/sumithrab/miniconda/envs/t2t/lib/python3.7/site-packages/gin/config.py", line 1055, in gin_wrapper
    return fn(*new_args, **new_kwargs)
  File "/home/sumithrab/miniconda/envs/t2t/lib/python3.7/site-packages/mesh_tensorflow/transformer/utils.py", line 2302, in run
    skip_seen_data=skip_seen_data)
  File "/home/sumithrab/miniconda/envs/t2t/lib/python3.7/site-packages/mesh_tensorflow/transformer/utils.py", line 1615, in train_m
odel
    estimator.train(input_fn=input_fn, max_steps=train_steps)
  File "/home/sumithrab/miniconda/envs/t2t/lib/python3.7/site-packages/tensorflow_estimator/python/estimator/tpu/tpu_estimator.py",
 line 3089, in train
    rendezvous.raise_errors()
  File "/home/sumithrab/miniconda/envs/t2t/lib/python3.7/site-packages/tensorflow_estimator/python/estimator/tpu/error_handling.py"
, line 150, in raise_errors
    six.reraise(typ, value, traceback)
  File "/home/sumithrab/miniconda/envs/t2t/lib/python3.7/site-packages/six.py", line 703, in reraise
    raise value
  File "/home/sumithrab/miniconda/envs/t2t/lib/python3.7/site-packages/tensorflow_estimator/python/estimator/tpu/tpu_estimator.py",
 line 3084, in train
    saving_listeners=saving_listeners)
  File "/home/sumithrab/miniconda/envs/t2t/lib/python3.7/site-packages/tensorflow_estimator/python/estimator/estimator.py", line 34
9, in train
    loss = self._train_model(input_fn, hooks, saving_listeners)
  File "/home/sumithrab/miniconda/envs/t2t/lib/python3.7/site-packages/tensorflow_estimator/python/estimator/estimator.py", line 11
75, in _train_model
    return self._train_model_default(input_fn, hooks, saving_listeners)
  File "/home/sumithrab/miniconda/envs/t2t/lib/python3.7/site-packages/tensorflow_estimator/python/estimator/estimator.py", line 12
04, in _train_model_default
    self.config)
  File "/home/sumithrab/miniconda/envs/t2t/lib/python3.7/site-packages/tensorflow_estimator/python/estimator/tpu/tpu_estimator.py",
 line 2921, in _call_model_fn
    config)
  File "/home/sumithrab/miniconda/envs/t2t/lib/python3.7/site-packages/tensorflow_estimator/python/estimator/estimator.py", line 1163, in _call_model_fn
    model_fn_results = self._model_fn(features=features, **kwargs)
  File "/home/sumithrab/miniconda/envs/t2t/lib/python3.7/site-packages/tensorflow_estimator/python/estimator/tpu/tpu_estimator.py", line 3212, in _model_fn
    _train_on_tpu_system(ctx, model_fn_wrapper, dequeue_fn))
  File "/home/sumithrab/miniconda/envs/t2t/lib/python3.7/site-packages/tensorflow_estimator/python/estimator/tpu/tpu_estimator.py", line 3654, in _train_on_tpu_system
    device_assignment=ctx.device_assignment)
  File "/home/sumithrab/miniconda/envs/t2t/lib/python3.7/site-packages/tensorflow/python/tpu/tpu.py", line 1665, in split_compile_and_shard
    name=name)
  File "/home/sumithrab/miniconda/envs/t2t/lib/python3.7/site-packages/tensorflow/python/tpu/tpu.py", line 1380, in split_compile_and_replicate
    outputs = computation(*computation_inputs)
  File "/home/sumithrab/miniconda/envs/t2t/lib/python3.7/site-packages/tensorflow_estimator/python/estimator/tpu/tpu_estimator.py", line 3640, in multi_tpu_train_steps_on_single_shard
    inputs=[0, _INITIAL_LOSS])
  File "/home/sumithrab/miniconda/envs/t2t/lib/python3.7/site-packages/tensorflow/python/tpu/training_loop.py", line 178, in while_loop
    condition_wrapper, body_wrapper, inputs, name="", parallel_iterations=1)
  File "/home/sumithrab/miniconda/envs/t2t/lib/python3.7/site-packages/tensorflow/python/ops/control_flow_ops.py", line 2774, in while_loop
    return_same_structure)
  File "/home/sumithrab/miniconda/envs/t2t/lib/python3.7/site-packages/tensorflow/python/ops/control_flow_ops.py", line 2256, in BuildLoop
    pred, body, original_loop_vars, loop_vars, shape_invariants)
  File "/home/sumithrab/miniconda/envs/t2t/lib/python3.7/site-packages/tensorflow/python/ops/control_flow_ops.py", line 2181, in _BuildLoop
    body_result = body(*packed_vars_for_body)
  File "/home/sumithrab/miniconda/envs/t2t/lib/python3.7/site-packages/tensorflow/python/tpu/training_loop.py", line 121, in body_wrapper
    outputs = body(*(inputs + dequeue_ops))
  File "/home/sumithrab/miniconda/envs/t2t/lib/python3.7/site-packages/tensorflow_estimator/python/estimator/tpu/tpu_estimator.py", line 3639, in <lambda>
    lambda i, loss: [i + 1, single_tpu_train_step(i)],
  File "/home/sumithrab/miniconda/envs/t2t/lib/python3.7/site-packages/tensorflow_estimator/python/estimator/tpu/tpu_estimator.py", line 1753, in train_step
    self._call_model_fn(features, labels))
  File "/home/sumithrab/miniconda/envs/t2t/lib/python3.7/site-packages/tensorflow_estimator/python/estimator/tpu/tpu_estimator.py", line 2031, in _call_model_fn
    estimator_spec = self._model_fn(features=features, **kwargs)
  File "/home/sumithrab/miniconda/envs/t2t/lib/python3.7/site-packages/mesh_tensorflow/transformer/utils.py", line 686, in my_model_fn
    log_file=model_info_file)
  File "/home/sumithrab/miniconda/envs/t2t/lib/python3.7/site-packages/mesh_tensorflow/ops.py", line 718, in __init__
    self.autostack()
  File "/home/sumithrab/miniconda/envs/t2t/lib/python3.7/site-packages/mesh_tensorflow/ops.py", line 836, in autostack
    max_combined_slice_size=max_combined_slice_size)
  File "/home/sumithrab/miniconda/envs/t2t/lib/python3.7/site-packages/mesh_tensorflow/ops.py", line 513, in rewrite_stack_variables
    slice_size = mesh_impl.slice_size(op.shape)
  File "/home/sumithrab/miniconda/envs/t2t/lib/python3.7/site-packages/mesh_tensorflow/ops.py", line 1004, in slice_size
    return list_product(self.slice_shape(tensor_shape))
  File "/home/sumithrab/miniconda/envs/t2t/lib/python3.7/site-packages/mesh_tensorflow/ops.py", line 977, in slice_shape
    % (tensor_shape, tensor_layout))
ValueError: Tensor dimension size not divisible by mesh dimension size: tensor_shape=Shape[heads=6, buckets=32] tensor_layout=TensorLayout(1, None)
  In call to configurable 'run' (<function run at 0x7fe77bf299e0>)
```

What are good values for these parameters and how do we get at them?
Your help is greatly appreciated.

@danyaljj FYI
[distillation_nqopen_xxl_ssm_to_small_ssm_operative_config.gin.txt](https://github.com/google-research/text-to-text-transfer-transformer/files/5519127/distillation_nqopen_xxl_ssm_to_small_ssm_operative_config.gin.txt)

**Comment by craffel** [comment]
Hi Sumithra, in the operative config you sent it looks like you had model_parallelism=1. Maybe it is not getting set to 8 appropriately? The OOM issues should be solvable via model parallelism and 8-way should be enough. Large batches should get serialized into microbatches, but you could also just try using a tiny batch size as a sanity check.

**Comment by sbhaktha** [comment]
Hi Colin,

The parameters in the `gin` file are overridden by the command line:

```python -m t5.models.mesh_transformer_main \
    --tpu=<v3_1024_tpu> \
    --gcp_project=<...> \
    --tpu_zone=<...> \
    --model_dir=<...> \
    --gin_file="dataset.gin" \
    --gin_file="distillation_nqopen_xxl_ssm_to_small_ssm_operative_config.gin" \
    --gin_param="MIXTURE_NAME = <...>" \
    --gin_param="utils.run.save_checkpoints_steps=<...>" \
    --gin_param="utils.run.batch_size=('tokens_per_batch', 393216)" \
    --gin_param="utils.run.train_steps=<...> \
    --gin_param="constant_learning_rate.learning_rate=<...>" \
    --gin_param="utils.tpu_mesh_shape.tpu_topology='16x32'" \
    --gin_param="utils.tpu_mesh_shape.model_parallelism=8"   \```

`model_parallelism` of 8 does not work and I get the above error (`ValueError: Tensor dimension size not divisible by mesh dimension size: tensor_shape=Shape[heads=6, buckets=32] tensor_layout=TensorLayout(1, None)`).

It seems to have to do with the number of heads being 6.

With `model_parallelism` set to 1 (either by explicitly setting it in the above command line or by not passing it here so that it uses the `1` in the `gin` file), I don't get this error. However I do run into the OOM error.

Another question is-- here we have two models- the teacher and the student and the model being fine-tuned is of course the student, which is only a `small` model. However we also need to have the teacher model (xxl) loaded and I think that is why we run into the OOM. Would `model_parallelism` be used for the bigger teacher model as well (not just the model being fine-tuned), because that is small enough anyway)?

**Comment by craffel** [comment]
Ah, I didn't look at the error you are getting, I just assumed it was an OOM. The issue is that the same model parallelism will be applied across the student and teacher models since all of their dimensions have the same name, and we can't divide the head dimension of the small model by 8 because it only has 6 heads. Is there any reason you are choosing the small model over a larger model to try distillation? If you are dead-set on a tiny model, you could try either renaming the dimensions in the small model so that mesh tensorflow doesn't try to distribute them, or you could pre-train a new small model that has dimensions that can be sliced 8 ways.

**Comment by sbhaktha** [comment]
Thanks Colin! We are trying to produce the smallest model possible via distillation and assess performance gains, but perhaps we could try one of the other models with a multiple of 8 heads. If we were to instead choose renaming the dimensions, how would we go about that? Is it just a change in the `gin` file or would it involve code changes as well?

**Comment by craffel** [comment]
I think you would need code changes in order to specify which dimension names to apply to which model (student vs. teacher).

I'm not sure whether initializing the student as a pre-trained model will be helpful or not. You could just copy the architecture for t5.1.1-small but make the # heads 8 instead of 6 and use that (randomly initialized, not pre-trained) as the student. That would still be a "small" model.

**Comment by sbhaktha** [comment]
Thank you!! How do we do random initialization (via the `gin` file)? Would we skip specifying an `init_checkpoint`?

**Comment by craffel** [comment]
Yes, just don't supply an init_checkpoint and it will randomly init.

**Comment by sbhaktha** [comment]
What exactly happens if we simply change the no. of heads to 8 but keep the `init_checkpoint` to point to the pretrained model? I am trying this and it seems to have loaded the student model successfully but I'd like to understand it better. Are any attention related parameters saved in the model? Does it technically make sense to load a pretrained model but change the no. of attention heads from the original model?

**Comment by craffel** [comment]
I think it will just split the parameters 8 ways into 6 ways. I don't think this really makes sense because, effectively, you will be chopping up weight matrices and recombining them in ways that they weren't originally used. But this may not be any worse than using random initialization. Who knows!

**Comment by sbhaktha** [comment]
So I fine-tuned an xxl_ssm model that was fine-tuned on the task we are interested in to a small model (8 heads,6 layers) for ~8000 steps and looked at the predictions and they look completely meaningless. It scores 0.78% on the 'exact match' metric we are using on the task. Sample inputs and outputs:

Inputs:
```
who sings does he love me with reba?'
how many pages is invisible man by ralph ellison?'
where do the great lakes meet the ocean?'
when does the new my hero academia movie come out?'
what is the smallest prime number that is greater than 30?'
```

Outputs:
```
The The The The The The - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
 ⁇  ''
the the United States
September 27 , 2017
 ⁇  ''
```

The teacher model was not great but scoring about. So clearly something is completely off. 

Also, when I try to evaluate using the `MtfModel.predict` API by pointing to the model path, it throws the following error. The `operatve_config.gin` in the model path is the distillation `gin` file  shared earlier, so it has all layer parameters contextualized under `teacher` or `student`. When the model is loaded for evaluation however, we are interested only in evaluating the (fine-tuned) student model, which is a `Bitransformer`.

```
 >> model_dir: gs://ai2-sumithrab/natural_questions_open_light_mixture/small_dis
tilled_from_xxl_ssm
 >> eval_split: validation
 >>>> writing model predictions . . .
Traceback (most recent call last):
  File "get_model_predictions.py", line 172, in <module>
    split=eval_split
  File "/home/sumithrab/miniconda/envs/t2t/lib/python3.7/site-packages/t5/models
/mtf_model.py", line 290, in eval
    vocabulary, score_in_predict_mode=eval_with_score)
  File "/home/sumithrab/miniconda/envs/t2t/lib/python3.7/site-packages/t5/models
/mtf_model.py", line 236, in estimator
    score_in_predict_mode=score_in_predict_mode)
  File "/home/sumithrab/miniconda/envs/t2t/lib/python3.7/site-packages/mesh_tens
orflow/transformer/utils.py", line 1529, in get_estimator
    mesh_shape=mesh_shape)
  File "/home/sumithrab/miniconda/envs/t2t/lib/python3.7/site-packages/mesh_tens
orflow/transformer/utils.py", line 196, in build_model
    layout=layout_rules)
  File "/home/sumithrab/miniconda/envs/t2t/lib/python3.7/site-packages/gin/confi
g.py", line 1078, in gin_wrapper
    utils.augment_exception_message_and_reraise(e, err_str)
  File "/home/sumithrab/miniconda/envs/t2t/lib/python3.7/site-packages/gin/utils
.py", line 49, in augment_exception_message_and_reraise
    six.raise_from(proxy.with_traceback(exception.__traceback__), None)
  File "<string>", line 3, in raise_from
  File "/home/sumithrab/miniconda/envs/t2t/lib/python3.7/site-packages/gin/confi
g.py", line 1055, in gin_wrapper
    return fn(*new_args, **new_kwargs)
  File "/home/sumithrab/miniconda/envs/t2t/lib/python3.7/site-packages/mesh_tens
orflow/transformer/transformer.py", line 1913, in make_bitransformer
    layer_stack=make_layer_stack(),
  File "/home/sumithrab/miniconda/envs/t2t/lib/python3.7/site-packages/gin/confi
g.py", line 1048, in gin_wrapper
    raise RuntimeError(err_str)
RuntimeError: Required bindings for `make_layer_stack` not provided in config: [
'layers']
  In call to configurable 'make_bitransformer' (<function make_bitransformer at 
0x7fd5fecdb710>)
```

If you have any further insight we'd appreciate it if you could share.

**Comment by craffel** [comment]
Hi Sumithra, what batch size are you using? You may just need a lot more training in order for the model to converge. Keep in mind that when using random initialization, the vocabulary matrix is totally meaningless, so the model needs to at least completely relearn language. It may be that distilling onto a non-pre-trained model is just inherently more difficult. I'd also point out that distilling from xxl to small is attempting to shrink by a factor of ~200; it's much more common to shrink by a factor of 2 or maybe 4 when distilling.

Regarding the gin error, it may be tricky to get it right - the studentteacher code is experimental and somewhat untested. I think the safest thing to do would be to manually remap the variables from the combined student/teacher checkpoint to a new checkpoint with appropriate variable names for a standalone Bitransformer model.

**Comment by sbhaktha** [comment]
Hi Colin, Thanks again for taking the time to reply! I forgot to mention in my comment that in this training I distilled from a fine-tuned xxl-ssm model to a standard small model with no. of heads = 8. And I used the standard small pretrained model as the `init_checkpoint`.

I think for now we are going to pause this direction and try other means of distillation. For now I will close this issue.

Appreciate your help!
